### PR TITLE
Workflow for ZPPY- PCMDI for E3SM diagnostics 

### PR DESCRIPTION
**Pull request by zhangshixuan1987**
## Issue resolution
- Closes #<ISSUE_NUMBER_HERE>

Select one: This pull request is...
- [ ] a bug fix: increment the patch version
- [ ] a small improvement: increment the minor version
- [ ] an incompatible (non-backwards compatible) API change: increment the major version

## 1. Does this do what we want it to do?

Objectives:
- Objective 1
- Objective 2
- ...
- Objective n

Required:
- [ ] Product Management: I have confirmed with the stakeholders that the objectives above are correct and complete.
- [ ] Testing: I have added or modified at least one "min-case" configuration file to test this change. Every objective above is represented in at least one cfg.
- [ ] Testing: I have considered likely and/or severe edge cases and have included them in testing.

If applicable:
- [ ] Testing: this pull request introduces an important feature or bug fix that we _must_ test often. I have updated the weekly-test configuration files, not just a "min-case" one.
- [ ] Testing: this pull request adds at least one new possible parameter to the cfg. I have tested using this parameter with and without any other parameter that may interact with it.

## 2. Are the implementation details accurate & efficient?

Required:
- [ ] Logic: I have visually inspected the entire pull request myself.
- [ ] Logic: I have left GitHub comments highlighting important pieces of code logic. I have had these code blocks reviewed by at least one other team member.

If applicable:
- [ ] Dependencies: This pull request introduces a new dependency. I have discussed this requirement with at least one other team member. The dependency is noted in `zppy/conda`, not just an `import` statement.

## 3. Is this well documented?

Required:
- [ ] Documentation: by looking at the docs, a new user could easily understand the functionality introduced by this pull request.

## 4. Is this code clean?

Required:
- [ ] Readability: The code is as simple as possible and well-commented, such that a new team member could understand what's happening.
- [ ] Pre-commit checks: All the pre-commits checks have passed.

If applicable:
- [ ] Software architecture: I have discussed relevant trade-offs in design decisions with at least one other team member. It is unlikely that this pull request will increase tech debt.

### CODE REVIEW

**Comment by forsyth2** [comment]
Thanks @zhangshixuan1987 for working on this! Let me know when it's ready for me to review. On my review, I can:
- Note code blocks that should be generalized/abstracted with parameters rather than hard-coded (e.g., how to take the code from prototype to production code).
- Confirm the checkboxes on the PR template above are fully met.
- Confirm the objectives / expected behavior, again noted on the PR template above.

**Comment by zhangshixuan1987** [comment]
@forsyth2 : Hi Ryan, based on what you mentioned above, I would like to ask you to review this commit first. This pull request have already included the key part for the workflow:
 1. the control script for job dependency: pcmdi_diags.py 
 2. the control template for the main calculation of pcmdi metrics:  pcmdi_diags.bash 
 3. the parameter files needed for certain process (template/pcmdi_diags/) 

I think that it would be better for you to provide some high-level comments on the overall structure and the coding style which should be first fulfilled to meet the requirment by Zppy. 

Thank you!

With best wishes,

Shixuan

File: zppy/templates/ts.bash [hunk 1]
```diff
@@ -147,19 +147,38 @@ EOF
...
   --var-list \
-  'mrsos, mrso, mrfso, mrros, mrro, prveg, evspsblveg, evspsblsoi, tran, tsl, lai, cLitter, cProduct, cSoilFast, cSoilMedium, cSoilSlow, fFire, fHarvest, cVeg, nbp, gpp, ra, rh' \
+  'snd, mrsos, mrso, mrfso, mrros, mrro, prveg, evspsblveg, evspsblsoi, tran, tsl, lai, cLitter, cProduct, cSoilFast, cSoilMedium, cSoilSlow, fFire, fHarvest, cVeg, nbp, gpp, ra, rh' \
   --realm \
   lnd \
   {% endif -%}
   {% if input_files.split(".")[0] == 'cam' or input_files.split(".")[0] == 'eam' -%}
   --var-list \
-  'pr, tas, rsds, rlds, rsus' \
+  'ua, va, ta, wa, zg, hur, pr, prc, prsn, ts, tas, prw, psl, sfcWind, tasmax, tasmin, tauu, tauv, rtmt, rsdt, rsds, rsdscs, rlds, rldscs, rsus, rsuscs, rsut, rsutcs, rlus, rlut, rlutcs, clivi, clwvi, clt, evspsbl, hfls, hfss, huss' \
```
Comment (forsyth2) [inline zppy/templates/ts.bash:181]: I believe we had some concern about explicitly stating variables in the `.bash` files rather than using the `cfg` to pass along variables. I see here however that we already had some hard-coded variables...

File: zppy/templates/pcmdi_diags.bash [hunk 2]
```diff
@@ -0,0 +1,1612 @@
...
+  exit 9
+fi
+
+# Copy output to web server
+echo
+echo ===== COPY FILES TO WEB SERVER =====
+echo
+
+# Create top-level directory
+web_dir=${www}/${case}/pcmdi_diags #/{{ sub }}
```
Comment (forsyth2) [inline zppy/templates/pcmdi_diags.bash:1550]: We had discussed, for production rather than prototyping, that we'd only put images on the `www` path, and keep everything else on the `output` path. This looks like it's copying everything over to the web server, is that right?

Comment (zhangshixuan1987) [reply zppy/templates/pcmdi_diags.bash:1486]: @forsyth2 : Yes, I have currently copied all output from PCMDI to the www path as I need to talk with you and Chengzhu to figure out the right sets of data that should be put in the www path.  Currently, I followed the E3SM_diags, as some of the diagnostic data from E3SM_diags will also be copied to www directory.

File: zppy/templates/ts.bash (hunk 1 above)
Comment (zhangshixuan1987) [reply zppy/templates/ts.bash:181]: This part is under-determined. I was trying not to break the original workflow for ts.bash, and it is better to discuss with you and Jill to come up a good solution or code changes to make the workflow more flexible.

File: zppy/templates/default.ini [hunk 3]
```diff
@@ -111,18 +113,210 @@ input_component = string(default="")
...
+# optional for mean climate (customized regional mean climate)
+regional = string(default="y")
+regions = string(default="global,ocean,land,NHEX,SHEX,TROPICS,NHEX_ocean,SHEX_ocean,NHEX_land,SHEX_land,ocean_50S50N")
+# optional for mean climate (save derived climatology data)
+save_test_clims = string(default=True)
+# option to process the mean climatology:
+climatology_process_method = string(default="nco")
+# Regridding by pcmdi (default is to regrid data to 2.5x2.5 grid for diagnostic metrics)
+# Required for mean climate
+# OPTIONS: '2.5x2.5' or an actual cdms2 grid object
```
[the remark this answers is not in the document]
Comment (zhangshixuan1987) [reply zppy/templates/default.ini:215]: @forsyth2 :   The command here passed a string rather than a command object to python. I think that we should have some online discussions for this.

File: zppy/templates/default.ini [hunk 4]
```diff
@@ -111,18 +113,210 @@ input_component = string(default="")
...
+# File of specified regions for mean climate calculation
+regions_specs = string(default="pcmdi_diags/regions_specs.json")
+# File of derived variables
+derived_variable = string(default="pcmdi_diags/derived_variable.json")
+# File of observation data name for mean climate calculation
+reference_alias = string(default="pcmdi_diags/reference_alias.json")
+# File of cmip6 metadata
+cmip_metadata = string(default="e3sm_to_cmip/default_metadata.json")
+# File of fixed pressulre level
+#cmip_plevdata = string(default="e3sm_to_cmip/vrt_remap_plev19.nc")
```
[the remark this answers is not in the document]
Comment (zhangshixuan1987) [reply zppy/templates/default.ini]: Yes, this becomes as redundent after I moved this part to "ts.bash".  I will follow you comments on this page and refine the code

File: zppy/pcmdi_diags.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
+import os
```
[the remark this answers is not in the document]
Comment (zhangshixuan1987) [reply zppy/pcmdi_diags.py:1]: Yes, this is adapted from the e3sm_diag.py.

File: zppy/templates/default.ini [hunk 6]
```diff
@@ -111,18 +113,210 @@ input_component = string(default="")
...
+  interp_vars = string(default=None)
 
 [tc_analysis]
 # NOTE: always overrides value in [default]
 input_files = string(default="eam.h2")
 # The scratch directory
 scratch = string(default="")
 
+[pcmdi_diags]
+# See url<need to work on document later>
```
[the remark this answers is not in the document]
Comment (zhangshixuan1987) [reply zppy/templates/default.ini:131]: I copied this from the e3sm_diags section, so I left these comments there to remind me that I should put a link here after I worked out the documentation.

File: zppy/templates/pcmdi_diags.bash [hunk 7]
```diff
@@ -0,0 +1,1612 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (zhangshixuan1987) [reply zppy/templates/pcmdi_diags.bash:1]: Yes, this is something that I would like to discuss with you and Chenzhu online.

**Comment by chengzhuzhang** [comment]
@zhangshixuan1987 and @forsyth2 I will begin reviewing this PR later next week. 
Shixuan, thank you for your hard work putting together this PR!

**Comment by zhangshixuan1987** [comment]
Hi @forsyth2 and @chengzhuzhang :  I am not sure if it is properate, but I want to share the sample .cfg files and the run output examples I created for the PCMDI here: 

-[ cfg file for v3.LR.hisotical simulation](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.szhang/e3sm-pcmdi/post.v3.LR.historical.cfg), which is adapted from Jill's cfg file 

- [sample output from the pcmdi](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.szhang/e3sm-pcmdi/v3.LR.historical_0051/pcmdi_diags/), which is applied to the v3.LR.historical simulations during 1985-2014 period. 

In addition, I also tested the workflow with AMIP and piControl simulations, which can work reasonably as well.  I need to further work on following things:
1.  verify the "model vs model" function 
2. discuss with you and Ryan for a. final www directory structure, and b. the viewer page.

File: zppy/pcmdi_diags.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (forsyth2) [reply zppy/pcmdi_diags.py:1]: @zhangshixuan1987 As part of #628, I did modularizing refactoring of `zppy` that greatly improves code readability/understandability + testing quality. We'll probably want to update this file to more closely match the other post-#628 `.py` files. I'm happy to help with that.

File: zppy/templates/ts.bash [hunk 8]
```diff
@@ -147,19 +147,38 @@ EOF
...
+  {% endif -%}
+
+  #call e3sm_to_cmip
   srun -N 1 e3sm_to_cmip \
   --output-path \
   ${dest_cmip}/${tmp_dir} \
   {% if input_files.split(".")[0] == 'clm2' or input_files.split(".")[0] == 'elm' -%}
   --var-list \
-  'mrsos, mrso, mrfso, mrros, mrro, prveg, evspsblveg, evspsblsoi, tran, tsl, lai, cLitter, cProduct, cSoilFast, cSoilMedium, cSoilSlow, fFire, fHarvest, cVeg, nbp, gpp, ra, rh' \
+  'snd, mrsos, mrso, mrfso, mrros, mrro, prveg, evspsblveg, evspsblsoi, tran, tsl, lai, cLitter, cProduct, cSoilFast, cSoilMedium, cSoilSlow, fFire, fHarvest, cVeg, nbp, gpp, ra, rh' \
```
Comment (chengzhuzhang) [inline zppy/templates/ts.bash:175]: `snd` doesn't seem like a variable `e3sm_to_cmip` support, maybe a typo?

File: zppy/templates/ts.bash (hunk 1 above)
Comment (chengzhuzhang) [reply zppy/templates/ts.bash:181]: @forsyth2 We talked about to have `e3sm_to_cmip` as a individual task, rather than embed it in `ts`, which will allow easier error handling, and expose more options to users to control. At this point, the hard-codded variables are all variables that `e3sm_to_cmip` support, if input for any variables are not available then they get skipped

**Review by chengzhuzhang — COMMENTED** [review]
@zhangshixuan1987 I just had chance to inspect the PR. It does seem like a call will work effectively to discussion design choice. I will schedule some time later next week. At the mean time, could you share a configuration file that I can try running by myself before the discussion.

**Comment by zhangshixuan1987** [comment]
@chengzhuzhang : Hi Jill, if you check my comments https://github.com/E3SM-Project/zppy/pull/624#issuecomment-2392406822. You will find the e3sm diag cfg file that I shared for you and Ryan to test. please take a look and let me know if you have any questions.

File: zppy/templates/ts.bash (hunk 8 above)
Comment (zhangshixuan1987) [reply zppy/templates/ts.bash:175]: snd is the snow depths defined in the CMIP6 model, but I am not sure if this variable is available at e3sm_to_cmip. By the way, this is a land model variable

**Comment by chengzhuzhang** [comment]
> Hi @forsyth2 and @chengzhuzhang : I am not sure if it is properate, but I want to share the sample .cfg files and the run output examples I created for the PCMDI here:
> 
>     * [ cfg file for v3.LR.hisotical simulation](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.szhang/e3sm-pcmdi/post.v3.LR.historical.cfg), which is adapted from Jill's cfg file
> 
>     * [sample output from the pcmdi](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.szhang/e3sm-pcmdi/v3.LR.historical_0051/pcmdi_diags/), which is applied to the v3.LR.historical simulations during 1985-2014 period.
> 
> 
> In addition, I also tested the workflow with AMIP and piControl simulations, which can work reasonably as well. I need to further work on following things:
> 
>     1. verify the "model vs model" function
> 
>     2. discuss with you and Ryan for a. final www directory structure, and b. the viewer page.

When reviewing the results, in the parallel coordinate figure: [example](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.szhang/e3sm-pcmdi/v3.LR.historical_0051/pcmdi_diags/model_vs_obs_1985-2014/graphics/mean_climate/e3sm/historical/v20241030/paracord_annual/rms_xyt_global_MODEL_VS_OBS_E3SM_HISTORICAL_V3-LR_mean_climate_ann_1985-2014.png). The `prw`, `rldscs`, `sfcWind` show large bias between v3, and CMIP MME, I think the legend might be mixed? Red line should be MME and black for v3?

When going over the results, I agree strongly that it would be much easier to navigate the results with a viewer page! @mingxuanwupnnl

**Comment by chengzhuzhang** [comment]
The `*.py` files under `template/pcmdi_diags ` are not installed correctly. 
Example:
`jinja2.exceptions.TemplateNotFound: pcmdi_diags/generate_sftlf.py

I think it is due to the constraint in setup:

https://github.com/E3SM-Project/zppy/blob/408e9ade87db8d0d5322c01186564df1392282be/setup.py#L20-L21

**Comment by zhangshixuan1987** [comment]
> > Hi @forsyth2 and @chengzhuzhang : I am not sure if it is properate, but I want to share the sample .cfg files and the run output examples I created for the PCMDI here:
> > ```
> > * [ cfg file for v3.LR.hisotical simulation](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.szhang/e3sm-pcmdi/post.v3.LR.historical.cfg), which is adapted from Jill's cfg file
> > 
> > * [sample output from the pcmdi](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.szhang/e3sm-pcmdi/v3.LR.historical_0051/pcmdi_diags/), which is applied to the v3.LR.historical simulations during 1985-2014 period.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In addition, I also tested the workflow with AMIP and piControl simulations, which can work reasonably as well. I need to further work on following things:
> > ```
> > 1. verify the "model vs model" function
> > 
> > 2. discuss with you and Ryan for a. final www directory structure, and b. the viewer page.
> > ```
> 
> When reviewing the results, in the parallel coordinate figure: [example](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.szhang/e3sm-pcmdi/v3.LR.historical_0051/pcmdi_diags/model_vs_obs_1985-2014/graphics/mean_climate/e3sm/historical/v20241030/paracord_annual/rms_xyt_global_MODEL_VS_OBS_E3SM_HISTORICAL_V3-LR_mean_climate_ann_1985-2014.png). The `prw`, `rldscs`, `sfcWind` show large bias between v3, and CMIP MME, I think the legend might be mixed? Red line should be MME and black for v3?
> 
> When going over the results, I agree strongly that it would be much easier to navigate the results with a viewer page! @mingxuanwupnnl

Hi @chengzhuzhang : I did not notice this message as you may @ the wrong people. As for your questions here:
1. Yes, the legend for CMIP MME and v3 in parallel coordinate plots was wrong, the fixed code with improvements in the mean climate metrics plots has been merged to my github branch. 
2. As for the  issues for `prw`, `rldscs`, `sfcWind`, this is partly related to the issues in the CMIP6 reference datasets. Jiwoo and I are working together to check the consistency of the units between two datasets, and will update this part.

**Comment by chengzhuzhang** [comment]
I accidentally tagged @mingxuanwupnnl, Sorry!

@zhangshixuan1987 sounds good, I will try test the new PR for now.
